readthedocs/api/v2/views/footer_views.py

```diff
@@ -63,113 +64,156 @@ def get_version_compare_data(project, base_version=None):
-    version = get_object_or_404(
-        Version.objects.public(
-            request.user,
-            project=project,
-            only_active=False,
-        ),
-        slug__iexact=version_slug,
-    )
-    main_project = project.main_language_project or project
+class FooterHTML(APIView):
 
-    if page_slug and page_slug != 'index':
-        if main_project.documentation_type == 'sphinx_htmldir':
-            path = page_slug + '/'
-        else:
-            path = page_slug + '.html'
-    else:
-        path = ''
-
-    version_compare_data = get_version_compare_data(project, version)
-
-    context = {
-        'project': project,
-        'version': version,
-        'path': path,
-        'downloads': version.get_downloads(pretty=True),
-        'current_version': version.verbose_name,
-        'versions': project.ordered_active_versions(user=request.user),
-        'main_project': main_project,
-        'translations': main_project.translations.all(),
-        'current_language': project.language,
-        'using_theme': using_theme,
-        'new_theme': new_theme,
-        'settings': settings,
-        'subproject': subproject,
-        'github_edit_url': version.get_github_url(
-            docroot,
-            page_slug,
-            source_suffix,
-            'edit',
-        ),
-        'github_view_url': version.get_github_url(
-            docroot,
-            page_slug,
-            source_suffix,
-            'view',
-        ),
-        'gitlab_edit_url': version.get_gitlab_url(
-            docroot,
-            page_slug,
-            source_suffix,
-            'edit',
-        ),
-        'gitlab_view_url': version.get_gitlab_url(
-            docroot,
-            page_slug,
-            source_suffix,
-            'view',
-        ),
-        'bitbucket_url': version.get_bitbucket_url(
-            docroot,
-            page_slug,
-            source_suffix,
-        ),
-        'theme': theme,
-    }
+    """
+    Render and return footer markup.
 
-    html = template_loader.get_template('restapi/footer.html').render(
-        context,
-        request,
-    )
-    resp_data = {
-        'html': html,
-        'show_version_warning': project.show_version_warning,
-        'version_active': version.active,
-        'version_compare': version_compare_data,
-        'version_supported': version.supported,
-    }
+    .. note::
 
-    # Allow folks to hook onto the footer response for various information
-    # collection, or to modify the resp_data.
-    footer_response.send(
-        sender=None,
-        request=request,
-        context=context,
-        resp_data=resp_data,
-    )
+       The methods `_get_project` and `_get_version`
+       are called many times, so a basic cache is implemented.
+    """
+
+    http_method_names = ['get']
+    permission_classes = [AllowAny]
+    renderer_classes = [JSONRenderer, JSONPRenderer]
+
+    def _get_project(self):
+        cache_key = '_cached_project'
```

Comment:
These cache keys should have a more specific prefix, like: `footerhtml_cached_project`. Otherwise, it's easy to override by mistake.

Reply:
It's an attribute on the class, so it's naturally scoped I think.

Reply:
Oh, right. We will need to change it if we move into a shared cache as you suggested, though